Asset Manager Service

davidbirchwork opened the pull request:
Asset Manager Service for OVE

senakafdo commented:
Tested service as well and it does seem to be working fine, but I need to test against the actual object store to be certain - will do tomorrow.

But, the API does seem slightly confusing. For instance I tried, `/OVEAssetModelController/GetWorkItem/{service}.{format}`. The docs say `Atomically get the next asset for working on for a given service Atomic update collisions will issue a 409 No Conflict, you should retry your request. spec if you want json or xml by appending .json or .xml`. 
I made a request with `Image` and `json` and I got a 204 response instead of the 200 response I expected. I then tried several other combinations and kept getting 409.

Also, there does seem to be a far for fundamental flaw with the API. It does seem to be expecting the user to provide XML or JSON as a format. This is usually dealt by introspecting the HTTP Accept header. Swagger injects this header by default and so does browsers but it seems to be ignored by the code.

davidbirchwork commented:
Thanks @senakafdo 

204 NoContent is expected functionality - it means that no unprocessed work items were found. I would certainly expect this if you made multiple requests to the service without uploading unprocessed content in the meantime. The other 409's would depend upon exactly what set of parameters you passed/

davidbirchwork commented:
My feedback on the Codacy report is as follows.

- The duplicate If statement is not duplicate and tests a condition potentially set two lines prior 

- Yes  the Sanatize method is a traditional use for a regex 

     -Here is one attempt at the full S3 bucket name regex from SO `(?=^.{3,63}$)(?!^(\d+\.)+\d+$)(^(([a-z0-9]|[a-z0-9][a-z0-9\-]*[a-z0-9])\.)*([a-z0-9]|[a-z0-9][a-z0-9\-]*[a-z0-9])$)`. I think we should avoid this as its unclear (at least to me!). Further, on behalf of the users,  I'd rather transform the string rather than  validate whether it meets a criteria, hence the string filtering approach rather than a Y/N regex. 

- I have fixed two trivial linting issues Codacy suggests 
- I don't see value in creating an exception class to be used in a single throw, hence reuse of System.Exception
- the issues regarding introducing new variables are non-necessary and would double the size of the method variable reuse would only cause problem if someone copied part of the code (which then would not compile)
 - i will read further on the configureawait method

Otherwise every other change suggested below would **break** the functionality of the API (an X below)
 
 
![image](https://user-images.githubusercontent.com/11289564/46788049-3c168300-cd31-11e8-9fcd-5eab0099d150.png)
![image](https://user-images.githubusercontent.com/11289564/46788061-42a4fa80-cd31-11e8-8509-83e1fc023b44.png)
![image](https://user-images.githubusercontent.com/11289564/46788067-46388180-cd31-11e8-9688-5ea10e4b1615.png)

davidbirchwork commented:
@senakafdo on the content type override, I implemented this way to avoid having one API method for returning a view and another for returning json/xml. I hadn't spotted this was overriding the content-type auto-negotiation. I've fixed this in the latest commit.

senakafdo commented:
@davidbirchwork, thanks for the feedback. 

Regarding the if statement, I have extracted the code below:

```csharp
if (ModelState.IsValid) {
      // then try and save it
      try {
        await _fileOperations.SaveFile(oveAssetModel, upload);

        _logger.LogInformation("received and uploaded a file :) " + oveAssetModel.StorageLocation);
    }
    catch (Exception e) {
        _logger.LogError(e, "failed to upload a file and write it to " + oveAssetModel.StorageLocation);
        ModelState.AddModelError("StorageLocation", "Failed to upload file");
    }
}

if (ModelState.IsValid) {
    _context.Add(oveAssetModel);
    await _context.SaveChangesAsync();
    return RedirectToAction(nameof(Index));
}
```
Did you intend to say that the ModelState is reset in the first block and the second has to then test it once again? If someone were refactoring the code, they may _honestly_ be tempted to combine these two blocks without even realising what it does.

Regarding the exception classes, perhaps we need to introduce a .net utils library, because though it is one occurrence for the Asset Manager service, there are few other occurrences for the Image Tile service.

Regarding the regex, what you would potentially do is provide a link to the [AWS page](https://docs.aws.amazon.com/AmazonS3/latest/dev/BucketRestrictions.html) and have a regex corresponding to that structure. I believe that the SO example is probably not accurate, [following the comments on that](https://stackoverflow.com/questions/50480924/regex-for-s3-bucket-name). You could probably have separate regular expressions testing the individual conditions which would make it much easier to read. Also, there seems to be better ways to construct a repeating group of letters and numbers compared to the SO regex you quoted.

In terms of the default values, can you explain how this URL would look under the following conditions:
`/OVEAssetModelController/ListAllAssets/{pos}/{count}.{format?}`:
1. If someone does not provide both pos and count
2. If someone provides pos but not count
3. If someone provides count but not pos
I'm having the feeling that the default values may not be required since those conditions may not be applicable anyway.

senakafdo commented:
> @senakafdo on the content type override, I implemented this way to avoid having one API method for returning a view and another for returning json/xml. I hadn't spotted this was overriding the content-type auto-negotiation. I've fixed this in the latest commit.

Cool, should we also be fixing the API urls in this case? Because you can similarly test for JSON vs XML without having this approach of passing `.json` or `.xml` as a part of the URL.

senakafdo commented:
@davidbirchwork, regarding codacy vs ReSharper - it uses Sonar for its C# analysis. I have read comparisons of the tools online and they seem to be equally popular. ReSharper is more popular in the developer community and Sonar with DevOps/automation - so I think we have the right balance here. If we feel that we are getting incorrect warnings, we can suppress them, [like you have done for ReSharper warnings](https://github.com/ove/ove-services/blob/bbbf3d3da4aba4eddedb708e381d15112e7d6ab0/packages/ove-asset-manager/src/OVE.Service.AssetManager/DbContexts/AssetModelContext.cs#L12).

davidbirchwork commented:
my point with code sonar is as demonstrated 92% of the issues found in this PR are not only misleading but would lead to broken code if implemented. the remaining 8% were linting. I do not find it helpful.

senakafdo commented:
> @senakafdo second check is required in the case an exception is thrown and ModelState.AddModelError("StorageLocation", "Failed to upload file"); is triggered. I think its pretty obvious that adding a model error makes the state invalid and thus requires the 2nd check. I can document this is you think necessary

Alright. 

I think this can be avoided altogether. So, `ModelState.IsValid` is false if an error is thrown right? If so, the following should do what you want isn't it?

```csharp
public async Task<ActionResult<OVEAssetModel>> Create(
    [Bind("Project,Name,Description,Service,AssetMeta")] OVEAssetModel oveAssetModel,
    [FromForm] IFormFile upload) {

    // check if we have a file
    if (upload == null || upload.Length <= 0) {
        _logger.LogError("failed to upload a file");
        ModelState.AddModelError("Filename", "Failed to upload file");
    } else if (!_serviceRepository.ValidateServiceChoice(oveAssetModel.Service, upload)) {
        ModelState.AddModelError("Service", "Service does not support File Type");
    } else {
        // then try and save it
        try {
            await _fileOperations.SaveFile(oveAssetModel, upload);

            _logger.LogInformation("received and uploaded a file :) " + oveAssetModel.StorageLocation);
        }
        catch (Exception e) {
            _logger.LogError(e, "failed to upload a file and write it to " + oveAssetModel.StorageLocation);
            ModelState.AddModelError("StorageLocation", "Failed to upload file");
        }
    }

    if (ModelState.IsValid) {
        _context.Add(oveAssetModel);
        await _context.SaveChangesAsync();
        return RedirectToAction(nameof(Index));
    }

    return this.FormatOrView(oveAssetModel);
}
```

davidbirchwork commented:
@senakafdo Chaining a set of independent tests like that together makes for very ugly looking and hard to maintain code. 
In general that code is not equivalent since it will not only return a single error back to the user rather than a set of errors. We should always seek to provide the most possible help to the user.

davidbirchwork commented:
for the exceptions, I see no value in writing a custom exception class and sharing between services purely so that it looks nicer in the logs and avoids a linter moaning, system.exception is a reasonable exception to throw

the .xml and .json are just syntactic sugar, helpful and easier to users, no one is obliged to supply them ;)

In terms of the default values, can you explain how this URL would look under the following conditions:
/OVEAssetModelController/ListAllAssets/{pos}/{count}.{format?}:
If someone does not provide both pos and count
values default to 0 and 100 
If someone provides pos but not count
count defaults to 100
If someone provides count but not pos
pos defaults to 0 
I'm having the feeling that the default values may not be required since those conditions may not be applicable anyway.
better to gracefully support users in their errors...

senakafdo commented:
> @senakafdo Chaining a set of independent tests like that together makes for very ugly looking and hard to maintain code.
> In general that code is not equivalent since it will not only return a single error back to the user rather than a set of errors. We should always seek to provide the most possible help to the user.

As a user I would rather expect the exact reason as to why my name is incorrect according to the criteria set forth by Amazon - but to help the user not make other mistakes WRT same name, the error message should ideally provide the URL such that they can read and find more information. This would be a much more user friendly approach compared to either of the logs `failed to sanitize s3 name`, `Failed to upload file for` followed by a failure to save the file.

Amazon may keep/change these rules at anytime, so I don't see anything easy/hard in terms of maintaining this implementation. Regular expressions are much better than naive string comparisons, which is what sets apart the existing approach from what is proposed.

davidbirchwork commented:
@senakafdo you are mixing up all of the issues in your last post. 

please can you skype me or speak in person, this is unproductive.

senakafdo commented:
> the .xml and .json are just syntactic sugar, helpful and easier to users, no one is obliged to supply them ;)

The format was mandatory on your Swagger output. Anyway, regardless of whether it is mandatory or not, this code is violating RESTful API design best practices. There are many resources to understand why this is not the best practice, such as [this](https://restfulapi.net/content-negotiation/).

BTW, what would the URLs actually look like (assume format == json)? I do understand the substitution of the default values and all. Is it \#1 or \#2 for instance when pos is missing?
1. /OVEAssetModelController/ListAllAssets/100.json
2. /OVEAssetModelController/ListAllAssets//100.json
Please note that your pos does not include the '/' of the URL. And, if your answer was \#1, how does the controller understand whether 100 was the pos or the count because they are both valid integers. This is what made me feel that these default values are never required. I see that the sonar warning is actually on point here.

jamesscottbrown commented:
> > the .xml and .json are just syntactic sugar, helpful and easier to users, no one is obliged to supply them ;)
> 
> The format was mandatory on your Swagger output. Anyway, regardless of whether it is mandatory or not, this code is violating RESTful API design best practices. There are many resources to understand why this is not the best practice, such as this.

I don't understand. The page that you link suggests two ways to do content negotiation in a RESTful API, and the second ("Content negotiation using URL patterns") seems to be what David is doing (returning a different format for the same content depending on whether the URL ends in .xml or .json).

senakafdo commented:
@jamesscottbrown, sorry, what I pointed out there, is that, there are two ways to test accepted content-type, and the approach defined in the RFC and accepted as the best practice is to use the Accept header. If you read the full article plus the RFC - you'll come to the same conclusion. The code uses the Accept header and the dot-format alternatively, which is typically what is suggested as a best to avoid.

Anyway, I've had an offline chat with David - we need to decide on whether Sonar checks are applicable for .net code or not. All of the above came up based on warning raised by Sonar (via codacy) and ways of resolving those. We find that we have been spending much longer debating about these, so along the lines of _spending time more productively_, we need to see if these checks are needed or not.

DavidAkroyd commented on packages/ove-asset-manager/src/OVE.Service.AssetManager/Views/OVEAssetModel/Create.cshtml:
```diff
@@ -0,0 +1,63 @@
+
+<h2>Create</h2>
+
+<h4>ImageFileModel</h4>
```
Please give this page a suitable name rather than ImageFileModel

davidbirchwork replied:
fixed

DavidAkroyd commented on packages/ove-asset-manager/src/OVE.Service.AssetManager/Views/OVEAssetModel/Create.cshtml:
```diff
@@ -0,0 +1,63 @@
+                <span asp-validation-for="Service" class="text-danger"></span>
+            </div>
+            <div class="form-group">
+                <label asp-for="AssetMeta" class="control-label"></label>
```
What is AssetMeta and how does it differ from Description?
I believe some more detail is required for these - I would recommend that all of the fields have an Input placeholder with a description of what is required

davidbirchwork replied:
AssetMeta is an optional place to store optional meta data about the asset which can be updated through API by other services. 

Have documented this.

DavidAkroyd commented on packages/ove-asset-manager/src/OVE.Service.AssetManager/Views/OVEAssetModel/Edit.cshtml:
```diff
@@ -0,0 +1,80 @@
+                <span asp-validation-for="AssetMeta" class="text-danger"></span>
+            </div>
+            <div class="form-group">
+                <label asp-for="StorageLocation" class="control-label"></label>
```
The Storage location should not be a form field, since a) it is read only, b) it should not be user changeable

In addition, if it is the actual location, this is a worthless field, since the unauthenticated user would have no access to a protected object store, so only a summary location would be required

davidbirchwork replied:
This feild is set to be readonly on line 45.

DavidAkroyd replied:
I think, as per other queries, this might be better as hidden, rather than just read-only?

DavidAkroyd commented on packages/ove-asset-manager/src/OVE.Service.AssetManager/Views/OVEAssetModel/Edit.cshtml:
```diff
@@ -0,0 +1,80 @@
+</div>
+
+<div>
+    <iframe src="@{ Services.GetViewURL(Model);}" onload="this.width=screen.width;this.height=1000px;" />
```
Why is this a hardcoded pixel size?

davidbirchwork replied:
fixed

DavidAkroyd commented on packages/ove-asset-manager/src/OVE.Service.AssetManager/Views/OVEAssetModel/GetRemovableView.cshtml:
```diff
@@ -0,0 +1,45 @@
+</div>
+
+<div>
+    <iframe src="@{ Services.GetViewURL(Model);}" onload="this.width=screen.width;this.height=1000px;" />
```
Again, hard coded pixel sizes are dangerous, especially on 4k screens(!)

davidbirchwork replied:
fixed

DavidAkroyd commented on packages/ove-asset-manager/src/OVE.Service.AssetManager/Views/Shared/Error.cshtml:
```diff
@@ -0,0 +1,22 @@
+﻿@model OVE.Service.AssetManager.Models.ErrorViewModel
```
Please can we update the Error page in order to provide a useful error for a production user

davidbirchwork replied:
This page only catches uncaught global exceptions which will make limited sense except to a developer. We could enable showing these exceptions on this page but it is not recommened: 

Swapping to Development environment will display more detailed information about the error that occurred.
Development environment should not be enabled in deployed applications as it can result in sensitive information from exceptions being displayed to end users.

DavidAkroyd commented on pm2-windows.json:
```diff
@@ -1,14 +1,25 @@
+      }
+    },
     {
       "name": "@ove/ove-service-imagetiles",
```
Is not Asset manager a replacement for ImageTiles rather than addition? If so, would this not break when you attempt to use PM2?

davidbirchwork replied:
ImageTileService will depend upon the Asset Manager Service

senakafdo replied:
@davidbirchwork, I raised an issue on the same point, without seeing this before. I believe @DavidAkroyd is pointing out that the pm2-windows.json does not seem to work based on how you have configured it.

davidbirchwork replied:
will fix in next commit

DavidAkroyd commented on pm2.json:
```diff
@@ -1,14 +1,24 @@
+          "PORT": "8181"
+      }
+    },{
       "name": "@ove/ove-service-imagetiles",
```
As above, this does not exist - does this not break?

davidbirchwork replied:
Image tile service and asset manager service will work together

DavidAkroyd commented on packages/ove-asset-manager/package.json:
```diff
@@ -0,0 +1,15 @@
+    "test": "echo \"Error: no test specified\" && exit 1",
+    "clean": "npx rimraf dist && npx rimraf src/OVE.Service.AssetManager/bin && npx rimraf src/OVE.Service.AssetManager/obj && npx rimraf AssetFiles.db",
+    "build": "dotnet restore src/OVE.Service.AssetManager && dotnet build src/OVE.Service.AssetManager && dotnet publish src/OVE.Service.AssetManager -o ../../dist",
+    "start": "dotnet dist/OVE.Service.AssetManager.dll --server.urls=http://0.0.0.0:${PORT-'8080'}"
```
You should remove default port of 8080, otherwise this may cause conflict with a missing config

DavidAkroyd replied:
For example, refusing to start if a port is not defined would be superior.
This would be ok if everything ran on its own container, however, since it is in the services, this can be presumed to not be the case

davidbirchwork replied:
removed

DavidAkroyd commented on packages/ove-asset-manager/src/OVE.Service.AssetManager/Domain/S3AssetFileOperations.cs:
```diff
@@ -0,0 +1,179 @@
+            _logger.LogInformation("about to upload "+asset);
+
+            // set up the filename            
+            asset.StorageLocation = Guid.NewGuid() +"/"+S3Santize(Path.GetFileNameWithoutExtension(upload.FileName))+Path.GetExtension(upload.FileName).Substring(0,4).ToLower();
```
So what happens if I upload a file.graphml ?
Is my Storage Location not then file.grap? (or on testing, .gra)

DavidAkroyd replied:
Why are you sanitising the filename? Only the bucket name requires sanitisation - the name can be any unicode
Some level of sanitisation would be required, however, as per the `Regex lowers` you do not need to remove starting digits from the filename, only the bucket

davidbirchwork commented on packages/ove-asset-manager/src/OVE.Service.AssetManager/Properties/launchSettings.json:
```diff
@@ -0,0 +1,27 @@
+      "commandName": "IISExpress",
+      "launchBrowser": true,
+      "environmentVariables": {
+        "ASPNETCORE_ENVIRONMENT": "Development"
```
These settings are only used during development, in docker this file is not used.

davidbirchwork commented:
Please note Remaining Codacy Issues are nonsensical as discussed above and their implementation would break the code. 
None of these show up in the Sonar C# plugin for Visual studio.

miguems commented:
Hi David,
why is making methods static non-sensical?

On Fri, Oct 12, 2018 at 8:09 AM David Birch <notifications@github.com>
wrote:

> Please note Remaining Codacy Issues are nonsensical as discussed above and
> their implementation would break the code.
> None of these show up in the Sonar C# plugin for Visual studio.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ove/ove-services/pull/7#issuecomment-429227415>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AEurBa3-JUJX5cO-agB7Qe1m236fyIKHks5ukEAkgaJpZM4XVkG->
> .
>

davidbirchwork commented:
Hi @miguems 
In general of course static methods are a good idea where they do not reference instance level variables. 
However the cases picked up by Codacy/Sonar here are erroneous because the are on an API controller. Only instance methods of the controller class will be picked up as API routes. Following the Codacy advice here would leave you with no API.  
The Linter I user (Resharper) is intelligent about this.

senakafdo commented on packages/ove-asset-manager/src/OVE.Service.AssetManager/GlobalSuppressions.cs:
```diff
@@ -0,0 +1,17 @@
+[assembly: System.Diagnostics.CodeAnalysis.SuppressMessage("Minor Code Smell", "S101:Types should be named in camel case", Justification = "Sonar has no custom dictionaries so cannot say OVE is ok", Scope = "type", Target = "~T:OVE.Service.AssetManager.Controllers.OVEAssetModelController")]
+[assembly: System.Diagnostics.CodeAnalysis.SuppressMessage("Major Code Smell", "S4144:Methods should not have identical implementations", Justification = "Sonar Does not understand how Views work", Scope = "member", Target = "~M:OVE.Service.AssetManager.Controllers.OVEAssetModelController.GetRemovableView(System.String)~System.Threading.Tasks.Task{Microsoft.AspNetCore.Mvc.ActionResult{OVE.Service ...
+[assembly: System.Diagnostics.CodeAnalysis.SuppressMessage("Minor Code Smell", "S1075:URIs should not be hardcoded", Justification = "its a single / ", Scope = "member", Target = "~M:OVE.Service.AssetManager.Domain.S3AssetFileOperations.ResolveFileUrl(OVE.Service.AssetManager.Models.OVEAssetModel)~System.String")]
+[assembly: System.Diagnostics.CodeAnalysis.SuppressMessage("Info Code Smell", "S1135:Track uses of \"TODO\" tags", Justification = "justification given on each occurance")]
```
@davidbirchwork, why is tracking TODO tags in code a false positive? I thought it makes sense to track these and warn the developer(s) that all of those needs to be removed before the code can be considered production-ready?

davidbirchwork replied:
rational was that unfortunately Sonar is not clever enough to pick up and report the rational from the code for each occurrence. 

I can remove this suppression if preferred.

senakafdo commented on packages/ove-asset-manager/src/OVE.Service.AssetManager/Domain/IAssetFileOperations.cs:
```diff
@@ -0,0 +1,13 @@
+using OVE.Service.AssetManager.Models;
+
+namespace OVE.Service.AssetManager.Domain {
+    public interface IAssetFileOperations {
```
Are these operations on Asset Files or Asset Models?

davidbirchwork replied:
These methods operate on the files of assets. 
please suggest a new name if preferred.

senakafdo commented on packages/ove-asset-manager/src/OVE.Service.AssetManager/Domain/IAssetFileOperations.cs:
```diff
@@ -0,0 +1,13 @@
+        string ResolveFileUrl(OVEAssetModel asset);
+        // ReSharper disable twice UnusedParameter.Global
+        Task<bool> MoveFile(OVEAssetModel oldAsset, OVEAssetModel newAsset);
+        Task<bool> DeleteFile(OVEAssetModel oveAssetModel);
```
I am sniffing - there seems to be smelly code here, :).

I think OVEAssetModel can/should have a better name. Why don't we call it OVEAsset? Why is the word Model required? Also, why are these methods ending with File? Is there a scenario where you delete a file and not the asset? It is slightly confusing to understand what's happening here.

davidbirchwork replied:
Following standard practice in asp.net is to append Model and Controller to such classes. 
I can break this practice if preferred? 

This class deals with files. The controller class deals with database. 
The controller class uses this class to deal with files in tandem. 

The controller class os decoupled from where the files are stored or interacting with them via this interface.

senakafdo commented:
I can do this kind of magic from the API. The processing state can become 100 apparently. Can we validate at API-level?

<img width="1124" alt="image" src="https://user-images.githubusercontent.com/3415383/46859714-6b4af400-ce06-11e8-80e2-97411ff1e959.png">

senakafdo commented on packages/ove-asset-manager/src/OVE.Service.AssetManager/GlobalSuppressions.cs:
```diff
@@ -0,0 +1,17 @@
+[assembly: System.Diagnostics.CodeAnalysis.SuppressMessage("Info Code Smell", "S1135:Track uses of \"TODO\" tags", Justification = "justification given on each occurance")]
+[assembly: System.Diagnostics.CodeAnalysis.SuppressMessage("Critical Code Smell", "S927:parameter names should match base declaration and other partial definitions", Justification = "tidier code", Scope = "member", Target = "~M:OVE.Service.AssetManager.Domain.S3AssetFileOperations.DeleteFile(OVE.Service.AssetManager.Models.OVEAssetModel)~System.Threading.Tasks.Task{System.Boolean}")]
+[assembly: System.Diagnostics.CodeAnalysis.SuppressMessage("Critical Code Smell", "S927:parameter names should match base declaration and other partial definitions", Justification = "tidier code", Scope = "member", Target = "~M:OVE.Service.AssetManager.Domain.S3AssetFileOperations.SaveFile(OVE.Service.AssetManager.Models.OVEAssetModel,Microsoft.AspNetCore.Http.IFormFile)~System.Threading.Tasks.Ta ...
+[assembly: System.Diagnostics.CodeAnalysis.SuppressMessage("Minor Code Smell", "S1075:URIs should not be hardcoded", Justification = "no other mechanism for specifying", Scope = "member", Target = "~M:OVE.Service.AssetManager.Startup.Configure(Microsoft.AspNetCore.Builder.IApplicationBuilder,Microsoft.AspNetCore.Hosting.IHostingEnvironment)")]
```
I believe that this encourages you to set the error page via a Constant or a Configuration file. I don't think the objective here is to tell you not to use a URL to define where to find the error page - I wouldn't consider the tool to be that silly, :).

davidbirchwork replied:
my thinking was simply that since you would never change the error page path independently of quite major code changes there is no purpose adding a configuration value.
I can add an extra loc to make this a constant if required but it seems a little like overkill as i should then extract every string in that class.

senakafdo replied:
ok if the .net convention was to always use that URL, then we can keep it as-is.

senakafdo commented on packages/ove-asset-manager/src/OVE.Service.AssetManager/GlobalSuppressions.cs:
```diff
@@ -0,0 +1,17 @@
+[assembly: System.Diagnostics.CodeAnalysis.SuppressMessage("Critical Code Smell", "S927:parameter names should match base declaration and other partial definitions", Justification = "tidier code", Scope = "member", Target = "~M:OVE.Service.AssetManager.Domain.S3AssetFileOperations.SaveFile(OVE.Service.AssetManager.Models.OVEAssetModel,Microsoft.AspNetCore.Http.IFormFile)~System.Threading.Tasks.Ta ...
+[assembly: System.Diagnostics.CodeAnalysis.SuppressMessage("Minor Code Smell", "S1075:URIs should not be hardcoded", Justification = "no other mechanism for specifying", Scope = "member", Target = "~M:OVE.Service.AssetManager.Startup.Configure(Microsoft.AspNetCore.Builder.IApplicationBuilder,Microsoft.AspNetCore.Hosting.IHostingEnvironment)")]
+[assembly: System.Diagnostics.CodeAnalysis.SuppressMessage("Minor Code Smell", "S101:Types should be named in camel case", Justification = "Sonar has no custom dictionaries so cannot say OVE is ok", Scope = "type", Target = "~T:OVE.Service.AssetManager.Models.OVEAssetModel")]
+[assembly: System.Diagnostics.CodeAnalysis.SuppressMessage("Major Code Smell", "S112:General exceptions should never be thrown", Justification = "not worth creating new exception class for 1 throw ", Scope = "member", Target = "~M:OVE.Service.AssetManager.Domain.S3AssetFileOperations.SaveFile(OVE.Service.AssetManager.Models.OVEAssetModel,Microsoft.AspNetCore.Http.IFormFile)~System.Threading.Tasks ...
```
Not a request to fix, but the above is a 50-50 case. If we are aiming to have more than one .Net service, better to have a utility library. Under ove-services, we can introduce a packages/ove-utils-assetmanagement or something. If this is the only service, then I have no issue as that would just be unwanted complexity.

davidbirchwork replied:
understood, however the danger is that we enforce .net for all services so will avoid doing this.

senakafdo replied:
Agreed.

senakafdo commented on packages/ove-asset-manager/src/OVE.Service.AssetManager/Migrations/20181009081536_InitialCreate.cs:
```diff
@@ -0,0 +1,38 @@
+﻿using System;
```
Is this class auto-generated? If not why `ReSharper disable All`?

davidbirchwork replied:
yes migrations are autogenerated

senakafdo replied:
I just realised that not all classes have the comment that says auto-generated.

senakafdo commented on packages/ove-asset-manager/src/OVE.Service.AssetManager/Views/Shared/_Layout.cshtml:
```diff
@@ -0,0 +1,71 @@
+    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
+    <title>@ViewData["Title"] - Data Science Institute</title>
+
+    <environment include="Development">
```
Is this a development-only file? I thought @DavidAkroyd did have a point re:Development raised earlier? Can we do an overall scan and get rid of this?

davidbirchwork replied:
no this is a production file. The line you are considering is the Razor equivilent of #if debug

senakafdo replied:
hmm, :). They should have used a better name, :). But don't they let you give your own names for the ASPNETCORE_ENVIRONMENT variable? This looks a bit ugly as-is.

senakafdo commented on packages/ove-asset-manager/src/OVE.Service.AssetManager/Views/Shared/_Layout.cshtml:
```diff
@@ -0,0 +1,71 @@
+        <link rel="stylesheet" href="~/lib/bootstrap/dist/css/bootstrap.css" />
+        <link rel="stylesheet" href="~/css/site.css" />
+    </environment>
+    <environment exclude="Development">
```
As above.

senakafdo commented on packages/ove-asset-manager/src/OVE.Service.AssetManager/Views/Shared/_Layout.cshtml:
```diff
@@ -0,0 +1,71 @@
+        <script src="~/lib/bootstrap/dist/js/bootstrap.js"></script>
+        <script src="~/js/site.js" asp-append-version="true"></script>
+    </environment>
+    <environment exclude="Development">
```
As above

davidbirchwork replied:
as above

senakafdo commented on packages/ove-asset-manager/src/OVE.Service.AssetManager/wwwroot/js/site.js:
```diff
@@ -0,0 +1 @@
+﻿// Write your site wide JavaScript code.
```
Is this file needed?

davidbirchwork replied:
not currently but I see little value in its removal. for example auth code would go here.

senakafdo replied:
Alternatively, we could also add that file at that point, where we need it, :).

senakafdo commented on pm2-windows.json:
```diff
@@ -1,14 +1,25 @@
+      "kill_timeout": 3000,
+      "restart_delay": 3000,
+      "script": "dotnet",
+      "args": "dist/OVE.Service.ImageTiles.dll --server.urls=http://0.0.0.0:8181",
```
Why does the asset manager refer to the ImageTiles dll? I don't quite understand how this actually manages to work. I think you are safe within the docker container since it only uses the pm2.json but not this one.

senakafdo commented on packages/ove-asset-manager/src/OVE.Service.AssetManager/Controllers/OVEAssetModelController.cs:
```diff
@@ -0,0 +1,517 @@
+            [FromForm] IFormFile upload) {
+
+            // check if we have a file
+            if (upload == null || upload.Length <= 0) {
```
@davidbirchwork, can you demonstrate how I can test the case of uploading a zero-byte file and getting this error? I have tried with an empty file (I cannot e-mail or attach since it is empty), and it doesn't show me any error.

<img width="460" alt="image" src="https://user-images.githubusercontent.com/3415383/46860979-6f2c4580-ce09-11e8-8040-e0b23e9a4fc1.png">

senakafdo commented on packages/ove-asset-manager/src/OVE.Service.AssetManager/Controllers/OVEAssetModelController.cs:
```diff
@@ -0,0 +1,517 @@
+                }
+                catch (DbUpdateConcurrencyException) {
+                    if (!_context.AssetModels.Any(e => e.Id == oveAssetModel.Id)) {
+                        return NotFound();
```
Isn't this a 500 error? compared to a 404 error?

davidbirchwork replied:
404 is NotFound, that is eactly the situation we have here, the specified asset ID does not exist in the database. 
If thats not the case the exception will be thrown as a 500.

senakafdo replied:
I think you get a [DbUpdateConcurrencyException](https://msdn.microsoft.com/en-us/library/system.data.entity.infrastructure.dbupdateconcurrencyexception(v=vs.113).aspx) when more than one person tries to perform the same operation and not when something is not found. 

Now what you are trying to do here is probably deal with a situation where another update changed the table and this record does not exist as a result. 

But, it may very well be the case that you deleted the old file but got the exception when trying to save the new file. If so the `404 - NotFound` status is wrong AFAIU. WDYT?

senakafdo replied:
@davidbirchwork, precisely - so not a 404 I suppose? I don't think that would mean anything to the API user. It would be a server-side error (5xx) too, because the user could have been impacted by another user.

davidbirchwork replied:
the if test  on line 458 establishes that no asset in the database exists for some reason. Therefore you can't edit it. a 404 seems logical to me. 

have changed the throw to return a Conflicted 409 HTTP status.

miguems commented:
> Hi @miguems
> In general of course static methods are a good idea where they do not reference instance level variables.
> However the cases picked up by Codacy/Sonar here are erroneous because the are on an API controller. Only instance methods of the controller class will be picked up as API routes. Following the Codacy advice here would leave you with no API.
> The Linter I user (Resharper) is intelligent about this.

I see... so for ASP.NET Web API to automatically create the API from the code, it forces that methods are instance public, regardless of their actual content or inner references. I find this a very stupid decision by MS if mandatory enforced, and a dangerous one just for the convenience of using wizards and autogen code...

senakafdo commented:
@oserban, @jamesscottbrown can we get more on hands on reviewing this please? I have been spending closer to a day on this PR and I think it takes a lot of effort to find all possible issues given the breadth of the code. I don't want to approve unless I'm certain that it has no issues. We need to help @davidbirchwork finish this off and then merge this PR.

senakafdo commented:
@davidbirchwork, as discussed offline, it seems that GitHub may delete comments added by others while a review is in progress. Can you kindly add your comments once again? I have them on e-mail, but better you add it compared to me pasting your comments. Thanks.

davidbirchwork commented:
@miguems this has nothing to do with autogen or wizards. Perhaphs a short tutorial on asp.net say [here](https://docs.microsoft.com/en-gb/aspnet/core/tutorials/first-web-api?view=aspnetcore-2.1) will help?  I would be happy to explain the logic of this rule or how asp.net core works to you at any time in person or lend you a book - just ask :-)

davidbirchwork commented on packages/ove-asset-manager/src/OVE.Service.AssetManager/Domain/IAssetFileOperations.cs:
```diff
@@ -0,0 +1,13 @@
+namespace OVE.Service.AssetManager.Domain {
+    public interface IAssetFileOperations {
+        string ResolveFileUrl(OVEAssetModel asset);
+        // ReSharper disable twice UnusedParameter.Global
```
R# is picking up that none of the implemenations of the interface use the two parameters. 
This is because the implemention of this method is marked as NotImplemented

davidbirchwork replied:
This is because moving files between s3 buckets is not recommended and is difficult to implement (e.g. amazon s3 does not provide API or SDK for it) and I'm uncertain it is something we want to support.

senakafdo replied:
Ok, it is strange that R# complains at the Interface and not at S3AssetFileOperations. Or probably it understands what `#pragma warning disable 1998` means.

davidbirchwork replied:
R# understands the NotImplementedException and so supresses the instance level message

davidbirchwork commented on packages/ove-asset-manager/src/OVE.Service.AssetManager/Domain/IAssetFileOperations.cs:
```diff
@@ -0,0 +1,13 @@
+        string ResolveFileUrl(OVEAssetModel asset);
+        // ReSharper disable twice UnusedParameter.Global
+        Task<bool> MoveFile(OVEAssetModel oldAsset, OVEAssetModel newAsset);
+        Task<bool> DeleteFile(OVEAssetModel oveAssetModel);
```
asp.net convention is to append Model, Controller, View to the name of the class for ease of differentiation. 
let me know if you want me to break this convention?

davidbirchwork replied:
Methods are named File to differentiate that this interface/class deals only with moving the Files of an Asset 

All methods relating to modifying Assets are handled by the API Controller which depends (via DI for decoupling) on an implementation of this interface.

senakafdo replied:
Ok so let me get this right. The asset model represents a file or files (I've noticed in some parts of the code) and then you would want to move that from A to B, delete it, etc. I didn't want to point out that `DeleteFiles` would be more appropriate, but I think you could for instance design this in a different way to avoid some of the complains from Sonar. 

For example, you are doing file operations on a file/files stored in S3, so you in fact could do have a Utility that gives you the list of files from an AssetModel and then run these operations on those files. But, as we have been discussing, these Sonar warnings are all about writing style of code not necessarily how it works. So, this is more of a preference rather than a necessity. The code may look better as a result, but this is something that is not mandatory to fix.

davidbirchwork replied:
The Asset model represents a single file.
Other servies may create many files (e.g. image tiles) but the Asset Service knows nothing about them.
The only instance multiple files are delt with is in the DeleteFile Method where every file in the s3 folder created for this asset are deleted. 

I will rename the `MoveFile` and `DeleteFile` method to make you/sonar happy.

davidbirchwork commented on packages/ove-asset-manager/src/OVE.Service.AssetManager/Domain/IAssetFileOperations.cs:
```diff
@@ -0,0 +1,13 @@
+using OVE.Service.AssetManager.Models;
+
+namespace OVE.Service.AssetManager.Domain {
+    public interface IAssetFileOperations {
```
They are operations on the Files of Assets

miguems commented:
> @miguems this has nothing to do with autogen or wizards. Perhaphs a short tutorial on asp.net say [here](https://docs.microsoft.com/en-gb/aspnet/core/tutorials/first-web-api?view=aspnetcore-2.1) will help? I would be happy to explain the logic of this rule or how asp.net core works to you at any time in person or lend you a book - just ask :-)

Yes, please. What is the logic of the rule? Because in the provided link there is nothing about static methods not being able to be part of an API.

davidbirchwork commented on packages/ove-asset-manager/src/OVE.Service.AssetManager/GlobalSuppressions.cs:
```diff
@@ -0,0 +1,17 @@
+[assembly: System.Diagnostics.CodeAnalysis.SuppressMessage("Minor Code Smell", "S101:Types should be named in camel case", Justification = "Sonar has no custom dictionaries so cannot say OVE is ok", Scope = "type", Target = "~T:OVE.Service.AssetManager.Controllers.OVEAssetModelController")]
+[assembly: System.Diagnostics.CodeAnalysis.SuppressMessage("Major Code Smell", "S4144:Methods should not have identical implementations", Justification = "Sonar Does not understand how Views work", Scope = "member", Target = "~M:OVE.Service.AssetManager.Controllers.OVEAssetModelController.GetRemovableView(System.String)~System.Threading.Tasks.Task{Microsoft.AspNetCore.Mvc.ActionResult{OVE.Service ...
+[assembly: System.Diagnostics.CodeAnalysis.SuppressMessage("Minor Code Smell", "S1075:URIs should not be hardcoded", Justification = "its a single / ", Scope = "member", Target = "~M:OVE.Service.AssetManager.Domain.S3AssetFileOperations.ResolveFileUrl(OVE.Service.AssetManager.Models.OVEAssetModel)~System.String")]
+[assembly: System.Diagnostics.CodeAnalysis.SuppressMessage("Info Code Smell", "S1135:Track uses of \"TODO\" tags", Justification = "justification given on each occurance")]
```
removed suppression, 

the examples in this codebase were already picked up by the linter as other issues, so was trying to avoid duplication.

davidbirchwork commented:
Hi @miguems 
Here are a few reasons that API methods must be instance level:
- Controllers are instanced once per request, this is to enable concurrency. 
- If you used a static method and accessed static state you would have the potential to run into write races, read races and other concurrency bugs. 
- Instance level methods are permitted to access request level state via the Controler 
- You cannot use dependency injection with a static methods model
- Without DI testing and mocking become much harder

DavidAkroyd commented:
@davidbirchwork I am happy with this from my QA point of view, though code will still need to be approved by Senaka or Ovidiu.
I would still request that the Storage Location field in the form be hidden rather than read-only since I don't think it presents any value to the user -it's only use would be through an API

senakafdo commented:
> @davidbirchwork I am happy with this from my QA point of view, though code will still need to be approved by Senaka or Ovidiu.
> I would still request that the Storage Location field in the form be hidden rather than read-only since I don't think it presents any value to the user -it's only use would be through an API

I will wait for Ovidiu and/or James to go through the code once before I approve - just so that it is vetted. I think in summary most of what we have been debating are non-functionality related issues. There are different ways of programming the same thing and these sometimes tend to be personal choices (like if-if-if vs if-elseif-elseif).

Tools like R#/Sonar are there to suggest what the common practice is, but if they often generate false positives. But, I think not every suggestion is a false-positive and somethings do present useful feedback with regards to writing high quality code. I have found that these have been addressed by @davidbirchwork.

Pending list of items would be to ensure there are no references to `Development` (if possible), a thorough review of the API, and to have proper documentation on the API and add some comments to make some doubtful parts of the code easy to read and understand. Re:API, I would suggest that @davidbirchwork checks the output of Swagger and confirms if he is happy with the URLs, the parameters that are required, and also the documentation that is in it.

For instance, `Atomically get the next asset for working on for a given service Atomic update collisions will issue a 409 No Conflict, you should retry your request. spec if you want json or xml by appending .json or .xml` probably is a result of a combination of different annotations/comments.

davidbirchwork commented:
@senakafdo Thank you, I believe commit ae269d8 resolves all of these issues

senakafdo commented:
@davidbirchwork, thanks for the fixes to the API. I have tried the following and I can confirm that they work as expected.
1. Uploaded an image
2. Edited the image
3. Deleted an image
4. Replaced an image with graph
those seem to be working w/o any issues.
Saw your changes re:API docs - looks better than before, thanks. There are a few things to polish and I look forward to you fixin these along with the new Image Tiles and Graph services.

I think there isn't much that I can add/change at this point by reviewing this further since I have provided all the feedback over a period of now closer to 2 days. So, unless @fguitton has anything to say specifically I have nothing further to request as changes.

I don't intend to be working further on this PR during the weekend or during the next week, so I'm approving this now. Please do wait for @fguitton, if you want his feedback or optionally go ahead with the merge as you requested.

davidbirchwork commented:
Thanks @senakafdo, do we really need to bother @fguitton ? This code has had a thorough review and approval by two team members.

davidbirchwork commented:
Thanks, I'm going to merge this one in now.
I'll happily fix any feedback from @fguitton or others in another (hopefully lets epic) PR :)

fguitton commented:
Well I guess the thing that comes up before even looking at the code is the lack of testing and documentation. The code looks clean and structured.

I have been looking at https://github.com/ove/ove-docs but could not find anything about what `assets`. I understand they are `objects` in a Object Store context. I am not sure what `processing state` it suppose to represent either, it is a serverless function result ?

In general I would centralise the content of your various discussions (in other issues) in a single place that helps contextualise. What is the point of having it in the asset service ? It is unclear what is the role or how to use the service from a developer perspective. You guys are developing your own jargon which will become quickly impossible to understand without a drawing or two ^^
